File: Lab 11/11-1.py
class Solution (object):
    def islandPerimeter(self,grid):
        perimeter = 0
        
        directions = [(-1,0), (1,0), (0,-1), (0,1)]
        
        for i in range (len(grid)):
            for j in range (len(grid[0])):
                if grid [i][j]==1:
                    cell_perimeter = 4
                    
                    for di, dj in directions:
                        ni, nj = i+di, j+dj
                        if 0 <= ni < len(grid) and 0 <= nj < len(grid[0]) and grid [ni][nj]==1:
                            cell_perimeter -=1
                            
                    perimeter += cell_perimeter
        return perimeter 

File: Lab 11/11-1.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int islandPerimeter(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == 1) {
                    queue.offer(new int[]{i, j});
                    visited[i][j] = true;
                    break;
                }
            }
            if (!queue.isEmpty()) break;
        }

        int perimeter = 0;

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0], col = cell[1];
            for (int[] dir : directions) {
                int newRow = row + dir[0];
                int newCol = col + dir[1];

                if (newRow < 0 || newRow >= rows || newCol < 0 || newCol >= cols || grid[newRow][newCol] == 0) {
                    perimeter++;
                } else if (!visited[newRow][newCol]) {
                    visited[newRow][newCol] = true;
                    queue.offer(new int[]{newRow, newCol});
                }
            }
        }
        return perimeter;
    }
}
